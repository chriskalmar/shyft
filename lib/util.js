
import DataType from './datatype/DataType';
import Entity from './entity/Entity';


export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : functionOrMap
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    typeof map === 'array' &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}

