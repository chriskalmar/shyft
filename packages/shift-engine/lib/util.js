import casual from 'casual';
import _ from 'lodash';



export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])



export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap }
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}


export const isString = (str) => (typeof str === 'string')


export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work')
  }

  return { ...first, ...second }
}



export const mapOverProperties = (object, iteratee) => {

  passOrThrow(
    isMap(object),
    () => 'Provided object is not a map'
  )

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function'
  )


  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[ key ], key)
  })
}



export const sortDataByKeys = (keys, data, keyProperty='id') => {
  const map = {}
  const result = []
  const order = {}

  if (!keys || ( isArray(keys) && keys.length === 0)) {
    return []
  }

  if (!data || ( isArray(data) && data.length === 0)) {
    return keys.map(() => null)
  }


  for (let d = 0; d < data.length; d++) {
    const row = data[ d ]
    const id = row[ keyProperty ]
    const found = map[id]

    if (!found) {
      map[id] = row
    }
    else if (found instanceof Array) {
      found.push(row)
    }
    else {
      map[id] = [ found, row ];
    }
  }


  for (let k = 0; k < keys.length; k++) {
    const key = keys[ k ]
    const found = map[ key ];

    if (typeof found === 'undefined') {
      result.push(null)
    }
    else if (found instanceof Array) {
      let idx = order[ key ];

      if (typeof idx === 'undefined') {
        idx = 0
      }
      else {
        idx++
      }

      order[ key ] = idx;

      result.push(found[ idx ]);
    }
    else {
      result.push(found);
    }
  }

  return result;
};



export const processCursor = (entity, cursor, orderBy, reverse) => {

  const $LT = reverse ? '$gt' : '$lt'
  const $GT = reverse ? '$lt' : '$gt'
  const $LTE = reverse ? '$gte' : '$lte'
  const $GTE = reverse ? '$lte' : '$gte'

  const where = {}

  if (cursor) {
    passOrThrow(
      isArray(cursor[ entity.name ]),
      () => 'Incompatible cursor for this entity'
    )

    passOrThrow(
      isArray(orderBy),
      () => 'orderBy needs to be an array of order definitions'
    )

    const orderList = []

    const orderMap = {}
    orderBy.map(({attribute, direction}) => {

      passOrThrow(
        attribute && direction,
        () => 'orderBy needs to be an array of attributes and respective sort order'
      )

      orderMap[ attribute ] = direction
    })

    const primaryAttribute = entity.getPrimaryAttribute()
    const attributes = entity.getAttributes()

    let foundUniqueAttribute = false

    cursor[ entity.name ].map(filter => {
      if (filter.length !== 2) {
        throw new Error('Cursor malformed')
      }

      const attributeName = filter[0]

      passOrThrow(
        attributes[ attributeName ],
        () => `Unknown attribute '${attributeName}' used in cursor`
      )

      if (attributeName !== primaryAttribute.name) {
        passOrThrow(
          orderMap[ attributeName ],
          () => `Cursor works only on sorted attributes (check: '${attributeName}')`
        )
      }

      const attribute = attributes[ attributeName ]

      // limit where clause to the first attribute which is defined as unique
      if (!foundUniqueAttribute) {
        orderList.push(attributeName)
      }

      if (attribute.isUnique) {
        foundUniqueAttribute = true
      }

    })

    passOrThrow(
      foundUniqueAttribute,
      () => 'Cursor needs to have at least one attribute defined as unique'
    )


    // if more than 2 attributes are used for the cursor take only the first and the last (primary key)
    if (orderList.length > 2) {
      orderList.splice(1, (orderList.length - 2) )
    }


    // simple filter for single attributes
    if (orderList.length === 1) {

      const attributeName = orderList[0]
      const value = cursor[ entity.name ][0][1]

      if (orderMap[ attributeName ] === 'DESC') {
        where[ attributeName ] = {
          [ $LT ]: value
        }
      }
      else {
        where[ attributeName ] = {
          [ $GT ]: value
        }
      }

    }
    else {
      where.$not = {}

      cursor[ entity.name ].map(filter => {
        const attributeName = filter[0]
        const value = filter[1]

        // ignore attributes that obsolete due to prior unique attribute
        if (orderList.indexOf(attributeName) === -1) {
          return
        }

        const attribute = attributes[ attributeName ]

        if (attribute.isUnique) {
          if (orderMap[ attributeName ] === 'DESC') {
            where.$not[ attributeName ] = {
              [ $GTE ]: value
            }
          }
          else {
            where.$not[ attributeName ] = {
              [ $LTE ]: value
            }
          }
        }
        else {
          where.$not[ attributeName ] = value

          if (orderMap[ attributeName ] === 'DESC') {
            where[ attributeName ] = {
              [ $LTE ]: value
            }
          }
          else {
            where[ attributeName ] = {
              [ $GTE ]: value
            }
          }
        }

      })
    }
  }

  return where
}


export const processCursors = (entity, args) => {

  const {
    after,
    before,
  } = args;

  const where = {
    $and: [
      processCursor(entity, after, args.orderBy),
      processCursor(entity, before, args.orderBy, true),
    ]
  }

  return where
}



export const reverseString = (str) => {
  return str
    .split('')
    .reverse()
    .join('');
}



export const randomJson = () => {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase(casual.words(2))
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer()

    ret[key] = value
  })

  return ret
}
