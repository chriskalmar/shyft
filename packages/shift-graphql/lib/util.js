
import { constants } from 'shift-engine';
import _ from 'lodash';


// generate a type name from a model
export function generateTypeName(entityModel) {

  if (!entityModel || !_.isObject(entityModel)) {
    throw new Error('generateTypeName() entityModel needs to be defined')
  }

  if (!entityModel.name || !entityModel.domain) {
    throw new Error('generateTypeName() entityModel needs a name and a domain')
  }

  let name = entityModel.domain

  // if it's not the local provider use it for the node name (unique naming)
  if (constants.localProviderName !== entityModel.provider) {
    name = `${entityModel.provider}_${name}`
  }

  // add entity name
  name += `_${entityModel.name}`

  return _.camelCase(name)
}


export function upperCaseFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}



export function generateTypeNameUpperCase(entityModel) {
  return upperCaseFirst(
    generateTypeName(entityModel)
  )
}


export default {
  generateTypeName,
  upperCaseFirst,
  generateTypeNameUpperCase,
}
