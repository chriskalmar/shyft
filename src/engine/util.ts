import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
