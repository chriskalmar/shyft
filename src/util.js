import crypto from 'crypto';

// postgres limitation
const SYSTEM_NAME_MAX_LENGTH = 63;

export const generateIndexName = (entityName, attributes, suffix = '_idx') => {
  let ret = `${entityName}_${suffix}_`;
  const attributesChain = (attributes || []).join('_');
  const attributesSortedChain = (attributes || []).sort().join('_');

  if (ret.length + attributesChain.length > SYSTEM_NAME_MAX_LENGTH) {
    const hashLength = 10;
    const hash = crypto
      .createHash('sha256')
      .update(attributesSortedChain)
      .digest('hex')
      .substr(0, hashLength);

    const shortendAttributesChain = attributesChain.substr(
      0,
      SYSTEM_NAME_MAX_LENGTH - hashLength - ret.length - 1,
    );

    ret = `${entityName}_${shortendAttributesChain}_${hash}_${suffix}`;
  } else {
    ret = `${entityName}_${attributesChain}_${suffix}`;
  }

  return ret;
};

export const invertDirection = direction => {
  return direction === 'DESC' ? 'ASC' : 'DESC';
};

export const getLimit = (first, last) => {
  if (first >= 0) {
    return first;
  } else if (last >= 0) {
    return last;
  }

  return 10;
};

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const quote = item => {
  if (!item) {
    throw new Error('quote() requires an input');
  }

  // special care for json pointers
  if (item.includes('->')) {
    const arr = item.split('->');
    arr[0] = `"${arr[0]}"`;
    return arr.join('->');
  }

  return `"${item}"`;
};
