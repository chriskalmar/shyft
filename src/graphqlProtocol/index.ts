import {
  getTypeForEntityFromGraphRegistry,
  extendModelsForGql,
  registerActions,
  generateGraphQLSchema,
} from './generator';
import {
  GraphQLJSON,
  GraphQLCursor,
  GraphQLBigInt,
  GraphQLDateTime,
  GraphQLDate,
} from './dataTypes';
import { ProtocolGraphQL } from './ProtocolGraphQL';
import {
  RELAY_TYPE_PROMOTER_FIELD,
  MAX_PAGE_SIZE,
} from './protocolGraphqlConstants';
import {
  ProtocolGraphQLConfiguration,
  isProtocolGraphQLConfiguration,
} from './ProtocolGraphQLConfiguration';
import { fromBase64, toBase64 } from './util';

export {
  getTypeForEntityFromGraphRegistry,
  extendModelsForGql,
  registerActions,
  generateGraphQLSchema,
  GraphQLJSON,
  GraphQLCursor,
  GraphQLBigInt,
  GraphQLDateTime,
  GraphQLDate,
  ProtocolGraphQL,
  RELAY_TYPE_PROMOTER_FIELD,
  MAX_PAGE_SIZE,
  ProtocolGraphQLConfiguration,
  isProtocolGraphQLConfiguration,
  fromBase64,
  toBase64,
};
