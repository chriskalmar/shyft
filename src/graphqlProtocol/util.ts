import * as _ from 'lodash';
import * as pluralize from 'pluralize';
import { RELAY_TYPE_PROMOTER_FIELD } from './protocolGraphqlConstants';
import { CustomError } from '../engine/CustomError';
import { isDefined } from '../engine/util';
import { getRegisteredEntityAttribute } from './registry';

export function generateTypeName(name) {
  return _.camelCase(name);
}

export function pascalCase(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function plural(str) {
  return pluralize.plural(str);
}

export function generateTypeNamePascalCase(name) {
  return pascalCase(generateTypeName(name));
}

export function generateTypeNamePlural(name) {
  return pluralize.plural(generateTypeName(name));
}

export function generateTypeNamePluralPascalCase(name) {
  return pascalCase(pluralize.plural(generateTypeName(name)));
}

export function generateTypeNameUpperCase(name) {
  return _.snakeCase(name).toUpperCase();
}

export const toBase64 = (value) =>
  new Buffer(value.toString()).toString('base64');
export const fromBase64 = (value) =>
  new Buffer(value.toString(), 'base64').toString();

export const serializeCursor = (cursor) => toBase64(JSON.stringify(cursor));
export const deserializeCursor = (cursor) => {
  try {
    return JSON.parse(fromBase64(cursor));
  } catch (e) {
    // TODO: check
    // should throw error, not return
    // https://github.com/graphql/graphql-js/issues/910
    return new Error('Invalid cursor provided');
  }
};

export const addRelayTypePromoterToInstance = (typeName, instance) => {
  if (!instance) {
    return instance;
  }

  instance[RELAY_TYPE_PROMOTER_FIELD] = typeName;
  return instance;
};

export const addRelayTypePromoterToList = (typeName, list) => {
  return list.map((instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  });
};

export const addRelayTypePromoterToInstanceFn = (typeName) => {
  return (instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  };
};

export const addRelayTypePromoterToListFn = (typeName) => {
  return (list) => {
    return addRelayTypePromoterToList(typeName, list);
  };
};

export const translateInstance = (
  entity,
  instance,
  { i18nLanguage, i18nDefaultLanguage, i18nLanguageDisableFallback },
) => {
  if (
    !instance ||
    !i18nLanguage ||
    i18nLanguage === i18nDefaultLanguage ||
    !entity.getI18nAttributeNames ||
    !entity.getI18nAttributeNames()
  ) {
    return instance;
  }

  const attributes = entity.getAttributes();
  const i18nAttributeNames = entity.getI18nAttributeNames();

  i18nAttributeNames.map((attributeName) => {
    const { required } = attributes[attributeName];
    const {
      fieldName: gqlFieldName,
      fieldNameI18n: gqlFieldNameI18n,
    } = getRegisteredEntityAttribute(
      entity.name,
      attributes[attributeName].name,
    );

    const translation = instance[gqlFieldNameI18n]
      ? instance[gqlFieldNameI18n][i18nLanguage]
      : undefined;

    if (i18nLanguageDisableFallback) {
      instance[gqlFieldName] =
        !isDefined(translation) && required
          ? new CustomError(
              `Translation for '${gqlFieldName}' not found`,
              'TranslationNotFoundError',
            )
          : translation;
    } else if (isDefined(translation)) {
      instance[gqlFieldName] = translation;
    }
  });

  return instance;
};

export const translateList = (entity, list, context) => {
  return list.map((instance) => {
    return translateInstance(entity, instance, context);
  });
};

export const translateInstanceFn = (entity, context) => {
  return (instance) => {
    return translateInstance(entity, instance, context);
  };
};

export const translateListFn = (entity, context) => {
  return (list) => {
    return translateList(entity, list, context);
  };
};

// emulate gql tag for prettier
export const gql = String.raw;
