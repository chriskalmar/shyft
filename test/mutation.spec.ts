/* eslint-disable dot-notation */
import './setupAndTearDown';
import { mutate, findOneByValue, testGraphql } from './db';

import {
  asUser,
  removeDynamicData,
  removeListDynamicData,
  removeId,
} from './testUtils';

import { BoardMember } from './models/BoardMember';
import { Board } from './models/Board';
import { Book } from './models/Book';
import { gql } from '../src/graphqlProtocol/util';

describe('mutation', () => {
  describe('via connector', () => {
    it('reject mutations without state transitions on stateful entities', async () => {
      const payload = {
        board: 50,
      };

      await mutate(BoardMember, 'create', payload, null, asUser(99)).catch(
        (e) => {
          expect(e).toMatchSnapshot();
        },
      );
    });

    it('perform create mutations', async () => {
      const payload = {
        board: 47,
      };

      const result = await mutate(
        BoardMember,
        'join',
        payload,
        null,
        asUser(99),
      );
      expect(removeDynamicData(BoardMember, result)).toMatchSnapshot();
    });

    it('perform create mutations with nested JSON attributes', async () => {
      const payload = {
        title: 'War and Peace',
        author: 'Leo Tolstoy',
        reviews: [
          {
            reviewer: 'John Connor',
            reviewText: "Couldn't stop reading",
            bookAttributes: [
              {
                attribute: 'Year of publishing',
                value: '1867',
              },
              {
                attribute: 'Pages',
                value: '1225',
              },
            ],
          },
        ],
      };

      const result = await mutate(Book, 'create', payload, null, asUser(99));
      expect(removeId(result)).toMatchSnapshot();
    });

    it('perform update mutations', async () => {
      const payload = {
        board: 50,
        invitee: 80,
      };

      let result = await mutate(
        BoardMember,
        'invite',
        payload,
        null,
        asUser(84),
      );
      const inviteId = result['id'];
      expect(removeDynamicData(BoardMember, result)).toMatchSnapshot();

      result = await mutate(BoardMember, 'accept', {}, inviteId, asUser(80));
      expect(removeDynamicData(BoardMember, result)).toMatchSnapshot();
    });

    it('perform update mutations with nested JSON attributes', async () => {
      const book = await findOneByValue(
        Book,
        { author: 'Leo Tolstoy' },
        asUser(99),
      );

      const payload = {
        reviews: [
          {
            reviewer: 'John Connor',
            reviewText: 'Updated review',
            bookAttributes: [
              {
                attribute: 'Year of publishing',
                value: '1867',
              },
              {
                attribute: 'Pages',
                value: '1225',
              },
              {
                attribute: 'Pages',
                value: '1225',
              },
              {
                attribute: 'Genre',
                value: 'Novel',
              },
            ],
          },
        ],
      };

      const result = await mutate(Book, 'update', payload, book.id, asUser(99));
      expect(removeId(result)).toMatchSnapshot();
    });

    it('handle uniqueness constraints', async () => {
      const payload = {
        name: 'New Board',
        isPrivate: false,
        vip: 1,
      };

      await mutate(Board, 'build', payload, null, asUser(99));

      await mutate(Board, 'build', payload, null, asUser(99)).catch((e) => {
        expect(e).toMatchSnapshot();
      });
    });

    it('perform delete mutations', async () => {
      const payload = {
        board: 50,
        invitee: 81,
      };

      let result = await mutate(
        BoardMember,
        'invite',
        payload,
        null,
        asUser(84),
      );
      const inviteId = result['id'];
      expect(removeDynamicData(BoardMember, result)).toMatchSnapshot();

      result = await mutate(BoardMember, 'remove', {}, inviteId, asUser(84));
      result['rows'] = removeListDynamicData(BoardMember, result['rows']);
      expect(removeDynamicData(BoardMember, result)).toMatchSnapshot();
    });

    it('fail mutation if fromState is not a match', async () => {
      const payload = {
        board: 50,
        invitee: 81,
      };

      const result = await mutate(
        BoardMember,
        'invite',
        payload,
        null,
        asUser(84),
      );
      const inviteId = result['id'];

      await mutate(BoardMember, 'accept', {}, inviteId, asUser(81));

      await mutate(BoardMember, 'accept', {}, inviteId, asUser(81)).catch(
        (e) => {
          expect(e).toMatchSnapshot();
        },
      );
    });

    it('reject mutations if ID not found', async () => {
      await mutate(BoardMember, 'remove', {}, 99999, asUser(80)).catch((e) => {
        expect(e).toMatchSnapshot();
      });
    });

    it('reject unknown mutations', async () => {
      await mutate(BoardMember, 'findInnerPeace', {}, 1, asUser(80)).catch(
        (e) => {
          expect(e).toMatchSnapshot();
        },
      );
    });
  });

  describe('via GrahpQL', () => {
    it('perform create mutations', async () => {
      const result = await testGraphql(
        gql`
          mutation joinBoardMember {
            joinBoardMember(input: { boardMember: { board: 47 } }) {
              boardMember {
                id
                board
                inviter
                invitee
                state
              }
            }
          }
        `,
        asUser(98),
      );

      expect(result).toMatchSnapshot();
    });

    it('perform create mutations with nested JSON attributes', async () => {
      const result = await testGraphql(
        gql`
          mutation createBook {
            createBook(
              input: {
                book: {
                  title: "War and Peace"
                  author: "Leo Tolstoy"
                  reviews: [
                    {
                      reviewer: "John Connor"
                      reviewText: "Couldn't stop reading"
                      bookAttributes: [
                        { attribute: "Year of publishing", value: "1867" }
                        { attribute: "Pages", value: "1225" }
                      ]
                    }
                  ]
                }
              }
            ) {
              book {
                id
                title
                author
                reviews {
                  reviewer
                  reviewText
                  bookAttributes {
                    attribute
                    value
                  }
                }
              }
            }
          }
        `,
        asUser(98),
      );

      expect(result).toMatchSnapshot();
    });

    it('perform update mutations', async () => {
      const inviteResult = await testGraphql(
        gql`
          mutation inviteBoardMember {
            inviteBoardMember(
              input: { boardMember: { board: 48, invitee: 70 } }
            ) {
              boardMember {
                id
                board
                inviter
                invitee
                state
              }
            }
          }
        `,
        asUser(41),
      );

      expect(inviteResult).toMatchSnapshot('invite');

      const inviteId = inviteResult.data.inviteBoardMember.boardMember.id;

      const acceptResult = await testGraphql(
        gql`
          mutation acceptBoardMemberById($inviteId: ID!) {
            acceptBoardMemberById(input: { id: $inviteId }) {
              boardMember {
                id
                board
                inviter
                invitee
                state
              }
            }
          }
        `,
        asUser(70),
        {
          inviteId,
        },
      );

      expect(acceptResult).toMatchSnapshot('accept');
    });
  });

  it('perform update mutations with nested JSON attributes', async () => {
    const book = await findOneByValue(
      Book,
      { author: 'Leo Tolstoy' },
      asUser(99),
    );

    const result = await testGraphql(
      gql`
        mutation updateBookById($id: ID!) {
          updateBookById(
            input: {
              id: $id
              book: {
                reviews: [
                  {
                    reviewer: "John Connor"
                    reviewText: "Updated review"
                    bookAttributes: [
                      { attribute: "Year of publishing", value: "1867" }
                      { attribute: "Pages", value: "1225" }
                      { attribute: "Pages", value: "1225" }
                      { attribute: "Genre", value: "Novel" }
                    ]
                  }
                ]
              }
            }
          ) {
            book {
              id
              title
              author
              reviews {
                reviewer
                reviewText
                bookAttributes {
                  attribute
                  value
                }
              }
            }
          }
        }
      `,
      asUser(98),
      {
        id: book.id,
      },
    );

    expect(result).toMatchSnapshot();
  });

  it('handle uniqueness constraints', async () => {
    const firstResult = await testGraphql(
      gql`
        mutation buildBoard {
          buildBoard(
            input: { board: { name: "Next Board", isPrivate: false, vip: 2 } }
          ) {
            board {
              id
            }
          }
        }
      `,
      asUser(98),
    );

    expect(firstResult).toMatchSnapshot('first');

    const secondResult = await testGraphql(
      gql`
        mutation buildBoard {
          buildBoard(
            input: { board: { name: "Next Board", isPrivate: false, vip: 2 } }
          ) {
            board {
              id
            }
          }
        }
      `,
      asUser(98),
    );

    expect(secondResult).toMatchSnapshot('second');
  });

  it('perform delete mutations', async () => {
    const inviteResult = await testGraphql(
      gql`
        mutation inviteBoardMember {
          inviteBoardMember(
            input: { boardMember: { board: 48, invitee: 75 } }
          ) {
            boardMember {
              id
              board
              inviter
              invitee
              state
            }
          }
        }
      `,
      asUser(41),
    );

    expect(inviteResult).toMatchSnapshot('invite');

    const inviteId = inviteResult.data.inviteBoardMember.boardMember.id;

    const removeResult = await testGraphql(
      gql`
        mutation removeBoardMemberById($inviteId: ID!) {
          removeBoardMemberById(input: { id: $inviteId }) {
            deleteRowCount
            id
          }
        }
      `,
      asUser(41),
      {
        inviteId,
      },
    );

    expect(removeResult).toMatchSnapshot('remove');
  });

  it('fail mutation if fromState is not a match', async () => {
    const inviteResult = await testGraphql(
      gql`
        mutation inviteBoardMember {
          inviteBoardMember(
            input: { boardMember: { board: 48, invitee: 72 } }
          ) {
            boardMember {
              id
              board
              inviter
              invitee
              state
            }
          }
        }
      `,
      asUser(41),
    );

    expect(inviteResult).toMatchSnapshot('invite');

    const inviteId = inviteResult.data.inviteBoardMember.boardMember.id;

    const firstAcceptResult = await testGraphql(
      gql`
        mutation acceptBoardMemberById($inviteId: ID!) {
          acceptBoardMemberById(input: { id: $inviteId }) {
            boardMember {
              id
              board
              inviter
              invitee
              state
            }
          }
        }
      `,
      asUser(72),
      {
        inviteId,
      },
    );

    expect(firstAcceptResult).toMatchSnapshot('firstAccept');

    const secondAcceptResult = await testGraphql(
      gql`
        mutation acceptBoardMemberById($inviteId: ID!) {
          acceptBoardMemberById(input: { id: $inviteId }) {
            boardMember {
              id
              board
              inviter
              invitee
              state
            }
          }
        }
      `,
      asUser(72),
      {
        inviteId,
      },
    );

    expect(secondAcceptResult).toMatchSnapshot('secondAccept');
  });

  it('reject mutations if ID not found', async () => {
    const result = await testGraphql(
      gql`
        mutation removeBoardMemberById($inviteId: ID!) {
          removeBoardMemberById(input: { id: $inviteId }) {
            deleteRowCount
            id
          }
        }
      `,
      asUser(41),
      {
        inviteId: 99999,
      },
    );

    expect(result).toMatchSnapshot();
  });
});
