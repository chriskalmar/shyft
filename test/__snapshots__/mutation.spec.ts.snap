// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation via connector fail mutation if fromState is not a match 1`] = `[Error: Instance with ID '304' not found or access denied]`;

exports[`mutation via connector handle uniqueness constraints 1`] = `[Error: Uniqueness constraint violated]`;

exports[`mutation via connector perform create mutations 1`] = `
Object {
  "board": "47",
  "id": "301",
  "invitee": "99",
  "inviter": "99",
  "state": 50,
}
`;

exports[`mutation via connector perform create mutations with nested JSON attributes 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "reviews": Array [
    Object {
      "bookAttributes": Array [
        Object {
          "attribute": "Year of publishing",
          "value": "1867",
        },
        Object {
          "attribute": "Pages",
          "value": "1225",
        },
      ],
      "reviewText": "Couldn't stop reading",
      "reviewer": "John Connor",
    },
  ],
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "en": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": undefined,
    "en": "War and Peace",
  },
}
`;

exports[`mutation via connector perform delete mutations 1`] = `
Object {
  "board": "50",
  "id": "303",
  "invitee": "81",
  "inviter": "84",
  "state": 10,
}
`;

exports[`mutation via connector perform delete mutations 2`] = `
Object {
  "deleteRowCount": 1,
  "id": "303",
  "rows": Array [
    Object {
      "board": "50",
      "id": "303",
      "invitee": "81",
      "inviter": "84",
      "state": 10,
    },
  ],
}
`;

exports[`mutation via connector perform update mutations 1`] = `
Object {
  "board": "50",
  "id": "302",
  "invitee": "80",
  "inviter": "84",
  "state": 10,
}
`;

exports[`mutation via connector perform update mutations 2`] = `
Object {
  "board": "50",
  "id": "302",
  "invitee": "80",
  "inviter": "84",
  "state": 20,
}
`;

exports[`mutation via connector perform update mutations with nested JSON attributes 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "reviews": Array [
    Object {
      "bookAttributes": Array [
        Object {
          "attribute": "Year of publishing",
          "value": "1867",
        },
        Object {
          "attribute": "Pages",
          "value": "1225",
        },
        Object {
          "attribute": "Pages",
          "value": "1225",
        },
        Object {
          "attribute": "Genre",
          "value": "Novel",
        },
      ],
      "reviewText": "Updated review",
      "reviewer": "John Connor",
    },
  ],
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "en": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": undefined,
    "en": "War and Peace",
  },
}
`;

exports[`mutation via connector reject mutations if ID not found 1`] = `[Error: Instance with ID '99999' not found or access denied]`;

exports[`mutation via connector reject mutations without state transitions on stateful entities 1`] = `[Error: State transition (toState) not defined]`;

exports[`mutation via connector reject unknown mutations 1`] = `[Error: Invalid mutation]`;
